casper/src/main/scala/coop/rchain/casper/BlockCreator.scala

```diff
@@ -65,19 +71,32 @@ object BlockCreator {
     spanF.trace(CreateBlockMetricsSource) {
       import cats.instances.list._
       for {
-        tipHashes             <- Estimator[F].tips(dag, genesis)
-        _                     <- spanF.mark("after-estimator")
-        parentMetadatas       <- EstimatorHelper.chooseNonConflicting(tipHashes, dag)
+        tipHashes <- Estimator[F].tips(dag, genesis)
+        _ <- Log[F].info(
+              s"Creating block with tip hashes ${tipHashes.map(PrettyPrinter.buildString)}"
+            )
+        _               <- spanF.mark("after-estimator")
+        parentMetadatas <- EstimatorHelper.chooseNonConflicting(tipHashes, dag)
+        _ <- Log[F].info(
+              s"Creating block with parents ${parentMetadatas.map(b => PrettyPrinter.buildString(b.blockHash))}"
+            )
         maxBlockNumber        = ProtoUtil.maxBlockNumberMetadata(parentMetadatas)
+        _                     <- Log[F].info(s"Creating block with maxBlockNumber ${maxBlockNumber}")
         invalidLatestMessages <- ProtoUtil.invalidLatestMessages(dag)
         deploys               <- extractDeploys(dag, parentMetadatas, maxBlockNumber, expirationThreshold)
         sender                = ByteString.copyFrom(validatorIdentity.publicKey.bytes)
         latestMessageOpt      <- dag.latestMessage(sender)
         seqNum                = latestMessageOpt.fold(0)(_.seqNum) + 1
+        _                     <- Log[F].info(s"Creating block with seqNum ${seqNum}")
+        invalidLatestMessagesExcludeUnbonded = invalidLatestMessages.filter(
+          v =>
+            parentMetadatas
+              .map(p => isBonded(p, v._1))
+              .forall(identity)
```

Comment:
As Greg said, we need to filter parents and take into account only ones that makes 67% of the whole stake.

Reply:
It can end up with several cases. How to deal with the cases propoerly?

Reply:
Instead of just returning boolean, `isBonded` can also return the stake. So here we can have overall stake behind each result. To decide if validator is bonded we don't need all of the parents (stake) to say yes, we only need 67% of the whole stake to say yes.
For synchrony constraint there is similar checking. Although it has a bug if only one validator is bonded.
https://github.com/rchain/rchain/blob/037411fc62cbb0115d6cb93fa2b7df62a320b649/casper/src/main/scala/coop/rchain/casper/SynchronyConstraintChecker.scala#L46-L58

Reply:
> we only need 67% of the whole stake to say yes.
Which `bondsMap` do we choose here?

Reply:
What I mean is we are leading a loop on choosing the right bondsMap here